### DISCUSSION TITLE
Graphviz works with minimal example, but not in Biosteam

### DISCUSSION THREAD

**Discussion by zorzansi** [BioSTEAMDevelopmentGroup/biosteam#158] [Q&A] [answered]
I have the following code from Biosteam tutorial. The last snippet from Graphviz works, so I assume it is correctly installed, but I do not get any diagram from Biosteam. Could you please help? Thank you.

```python
import biosteam as bst
import os
from biosteam import settings
path_variable = os.environ.get('PATH')
print(path_variable)
bst.nbtutorial() # Light-mode html diagrams and filter warnings
settings.set_thermo(['Water', 'Methanol'])
feed = bst.Stream(Water=50, Methanol=20)
feed.show()

feed.price = 0.15 # USD/kg
feed.cost # USD/hr
settings.CEPCI # Default year is 2017
settings.CEPCI = 603.1 # To year 2018

settings.electricity_price # Default price (USD/kWhr)

settings.electricity_price = 0.065 # Adjust price
settings.cooling_agents # All available cooling agents

cooling_water = settings.get_cooling_agent('cooling_water')
cooling_water.show() # A UtilityAgent

# Price of regenerating the utility in USD/kmol
cooling_water.regeneration_price

# Other utilities may be priced for amount of heat transfered in USD/kJ
chilled_water = settings.get_cooling_agent('chilled_water')
chilled_water.heat_transfer_price

cooling_water.T = 302 # Change the temperature of cooling water (K)

settings.heating_agents # All available heating agents

lps = settings.get_heating_agent('low_pressure_steam') # A UtilityAgent
lps.show() # Note that because utility changes phase, T_limit is None

lps.regeneration_price = 0.20 # Adjust price (USD/kmol)

from biosteam import units

# Specify vapor fraction and isobaric conditions
F1 = units.Flash('F1', V=0.1, P=101325)
F1.show()

F1.ins

F1.outs

F1.ins[0] = feed
F1.show()

F1.simulate()
F1.show()

print(F1.results()) # Default returns DataFrame object with units

F1.net_duty # Duty [kJ / hr]

F1.net_power # Electricity consumption [kW]

[F1.feed, F1.vapor, F1.liquid] # Inlet feed and vapor and liquid outlets

M1 = units.Mixer('M1')
S1 = units.Splitter('S1', outs=('liquid_recycle', 'liquid_product'),
                    split=0.5) # Split to 0th output stream
F1.outs[0].ID = 'vapor_product'
F1.outs[1].ID = 'liquid'

bst.main_flowsheet.diagram()
# Note that empty streams are dashed and the 
# width of streams depend on their flow rates (by mass)

feed = bst.Stream('feed', Methanol=100, Water=450)

# Broken down -pipe- notation
[S1-0, feed]-M1     # M1.ins[:] = [S1.outs[0], feed]
M1-F1               # F1.ins[:] = M1.outs
F1-1-S1             # S1.ins[:] = [F1.outs[1]]

# All together
[S1-0, feed]-M1-F1-1-S1;

#NONE OF THIS WORKS
bst.main_flowsheet.diagram() 
bst.F.diagram() 

# THIS WORKS

import graphviz
g = graphviz.Digraph('G', filename='hello.gv')
g.edge ('Hello', 'World')
g.view()
```

**Comment by yalinli2** [comment]
Are there objectives on your `main_flowsheet`?  What will `bst.main_flowsheet.registries` give you?

Can you try an established system:
```python
from biorefineries import cornstover as cs
cs.load()
cs.cornstover_sys.diagram()
```

If you can see a figure, then everything is fine, if there are problems with graphviz, you should get a warning message mentioning graphviz installation issues, and you may get more information by firstly letting biosteam raise the error:
```python
bst.preferences.raise_exception = True
# then repeat whatever needed to reproduce the problem, will be super helpful if you post the warning/error messages with full traceback
```

**Reply by zorzansi** [reply]
Dear Yalin,
I tried the following code:
```
from biorefineries import cornstover as cs
import biosteam as bst
import graphviz
g = graphviz.Digraph('G', filename='hello.gv')
g.edge('Hello', 'World')
g.view()
bst.preferences.raise_exception = True
cs.load()
foo=cs.cornstover_sys.diagram()
print(foo)
```

I confirm Graphviz is working (I get the diagram in pdf and in the browser), while foo is NoneType: None

I am getting some warnings, I am running on an environment with Python 3.10

> C:\Anaconda3\envs\py310\lib\site-packages\biorefineries\cellulosic\systems\cellulosic_ethanol.py:134: RuntimeWarning: cannot define both moisture split and moisture content
>   S401 = bst.PressureFilter('S401', (stillage, recycled_water))
> C:\Anaconda3\envs\py310\lib\site-packages\thermosteam\equilibrium\bubble_point.py:103: RuntimeWarning: NH3 has no defined Dortmund groups; functional group interactions are ignored
>   self.gamma = thermo.Gamma(chemicals)
> C:\Anaconda3\envs\py310\lib\site-packages\thermosteam\equilibrium\dew_point.py:112: RuntimeWarning: NH3 has no defined Dortmund groups; functional group interactions are ignored
>   self.gamma = thermo.Gamma(chemicals)
> C:\Anaconda3\envs\py310\lib\site-packages\biosteam\units\design_tools\tank_design.py:107: CostWarning: volume (29876 gal) is below the lower bound (30000 gal) for purchase cost estimation
>   warn(f"volume ({V_total:.5g} {V_units}) is below "

  
Re-running the code of my first message and adding the line
`bst.preferences.raise_exception = True`
just after the import, does not rise any exception.

Any idea? 
Thank you!

**Reply by yalinli2** [reply]
Hmmm unfortunately I've never run into this before, we did have a colleague who was using PyCharm and had problems getting the diagrams... When using Spyder I always have the diagrams in the plot window (not the browser) without a problem, tagging @yoelcortes and see if he has any ideas

**Reply by zorzansi** [reply]
I got the reply from @yoelcortes and indeed, by using a file and then its view function, I can get the diagram:

This doest not produce a diagram:

```
bst.main_flowsheet.diagram()
```
This does:
```
f=bst.main_flowsheet.diagram(display=False)
f.view('filename')
```

Still, it would be nice to fix the issue, to follow more easily the beginners guide.

**Reply by yoelcortes** [reply]
@zorzansi, I'll add a note that users need to use the IPython console to display diagrams. Using the "view" command by default within BioSTEAM is not a good solution because it saves unwanted files and opens it externally (so it is harder to navigate), but I can also make it an option. 

Note that this not an issue with BioSTEAM. It works as intended out of the box in Jupyter notebook and Spyder (with graphviz installed). BioSTEAM also includes IPython>=7.9.0 as a requirement in the installation, so PyCharm users just need to adjust their settings to use the IPython console. I would recommend submitting a question with PyCharm or on stackexchange if you are having trouble with this. 

Thanks!

**Comment by yoelcortes** [comment, marked as answer]
@zorzansi, @yalinli2 

Yeah, @Scheideba used PyCharm and had this same problem. The issue was he was running his code through the terminal and not through the ipython console. If this is the issue, maybe the following thread can help:
https://stackoverflow.com/questions/47765237/cant-use-ipython-console-in-pycharm

Alternatively, if you prefer to use the terminal (not ipython), you can view diagrams through the graphviz view command (which saves a file and then opens it externally) as follows:

```python
from biorefineries import cornstover as cs
cs.load()
f = cs.cornstover_sys.diagram(
    # Do not display in IPython console and return a graphviz.graphs.Digraph object
    display=False
)
f.view('filename')
```

For additional details on making diagrams (such as saving them directly), check out https://biosteam.readthedocs.io/en/latest/API/System.html#biosteam.System.diagram

I hope this helps,
Thanks!
